src/serialize.h

```diff
@@ -109,6 +109,18 @@ static uint16_t read_uint16(ios_t *s) JL_NOTSAFEPOINT
     return x;
 }
 
+static void write_uint32(ios_t *s, uint32_t i) JL_NOTSAFEPOINT
+{
+    ios_write(s, (char*)&i, 4);
+}
+
+static uint32_t read_uint32(ios_t *s) JL_NOTSAFEPOINT
+{
+    uint32_t x = 0;
+    ios_read(s, (char*)&x, 4);
+    return x;
+}
+
```

Comment:
These functions are duplicates of https://github.com/JuliaLang/julia/blob/c50361144ce783cecf65294e09d6e7cfdc73eb3f/src/staticdata.c#L371-L381.  Also, introducing them introduced a new warning when compiling julia: this header file is included by `ircode.c` which doesn't use these functions.  I'm not even sure where there are functions implementations in a header.

Reply:
What is the best way in C to define utility functions in a single place and import them from several places? In this case, dump.c uses all `write_xxx` utilities defined in serialize.h while ircode.c only uses some of them and so `#include serialize.h` produced that warning (and the same thing would happen for staticdata.c if we remove the duplications).

Reply:
> What is the best way in C to define utility functions in a single place and import them from several places?

Move the implementations (not the prototypes!) to a `.c` file, and make the files which need those functions depend on its corresponding object file in `src/Makefile`

Reply:
Isn't the reason it is done like this to make the function definition available in all translation units and thus be eligible for inlining? But these functions are not marked `inline` though... But, to me, for consistency, the function in `staticdata.c` should be removed and this header included there instead?

Reply:
#44739